### PR TITLE
support for CRD resource requirements

### PR DESCRIPTION
**Pull request by zolug**
- Resource Requirements can be set for Trench, Attractor and Conduit through annotation. (key=resource-template)
- For each container making up a specific CRD the annotation value is interpreted as the name of a Resource Requirements template. (So templates are defined per container.)
- CRD annotation update is supported and will be applied according to the related deployment/daemonset/statefulset 'Update Strategy'.
- The Resource Requirement templates are stored in a Config Map, and its content is mounted as a volume in the Operator POD.
- The Config Map can be updated on the fly if needed. But changes do not take effect on existing CRDs, unless the associated Reconcile() function is invoked by kubernetes for some reason.
 - A template basically wraps a k8s corev1 API ResourceRequirements.
 - Currently there are 3 predefined Resource Requirements templates per container with names: "small", "medium", "large".
 
 Example;
    
      cat <<EOF | kubectl apply -f -
      apiVersion: meridio.nordix.org/v1alpha1
      kind: Trench
      metadata:
        name: trench-a
        annotations:
          resource-template: "small"
      spec:
        ip-family: dualstack
      EOF

### CODE REVIEW

File: controllers/attractor/nse-vlan.go [hunk 1]
```diff
@@ -114,6 +120,11 @@ func (i *NseDeployment) insertParameters(dep *appsv1.Deployment) *appsv1.Deploym
 				container.Ports = append([]corev1.ContainerPort{}, corev1.ContainerPort{HostPort: common.VlanNsePort, ContainerPort: common.VlanNsePort})
 			}
 			container.Env = i.getEnvVars(container.Env)
+			// set resource requirements for container (if not found, then values from model
+			// are kept even upon updates, as getReconciledDesiredStatus() overwrites containers)
+			if err := common.SetContainerResourceRequirements(&i.attractor.ObjectMeta, &container); err != nil {
+				i.exec.LogInfo(fmt.Sprintf("attractor %s, %v", i.attractor.ObjectMeta.Name, err))
```
Comment (LionelJouin) [inline controllers/attractor/nse-vlan.go:126]: LogError would be more appropriate

File: controllers/attractor/lb-fe.go [hunk 2]
```diff
@@ -189,12 +195,22 @@ func (l *LoadBalancer) insertParameters(dep *appsv1.Deployment) *appsv1.Deployme
 					common.GetProbeCommand(false, "unix:///tmp/health.sock", ""))
 			}
 			container.Env = l.getLbEnvVars(container.Env)
+			// set resource requirements for container (if not found, then values from model
+			// are kept even upon updates, as getReconciledDesiredStatus() overwrites containers)
+			if err := common.SetContainerResourceRequirements(&l.attractor.ObjectMeta, &container); err != nil {
+				l.exec.LogInfo(fmt.Sprintf("attractor %s, %v", l.attractor.ObjectMeta.Name, err))
```
Comment (LionelJouin) [inline controllers/attractor/lb-fe.go:201]: LogError

File: controllers/attractor/lb-fe.go [hunk 3]
```diff
@@ -189,12 +195,22 @@ func (l *LoadBalancer) insertParameters(dep *appsv1.Deployment) *appsv1.Deployme
...
 		case "nsc":
 			if container.Image == "" {
 				container.Image = "registry.nordix.org/cloud-native/nsm/cmd-nsc:v1.3.0"
 				container.ImagePullPolicy = corev1.PullAlways
 			}
 			container.Env = l.getNscEnvVars(container.Env)
+			// set resource requirements for container (if not found, then values from model
+			// are kept even upon updates, as getReconciledDesiredStatus() overwrites containers)
+			if err := common.SetContainerResourceRequirements(&l.attractor.ObjectMeta, &container); err != nil {
+				l.exec.LogInfo(fmt.Sprintf("attractor %s, %v", l.attractor.ObjectMeta.Name, err))
```
Comment (LionelJouin) [inline controllers/attractor/lb-fe.go:212]: LogError

File: controllers/attractor/lb-fe.go [hunk 4]
```diff
@@ -213,6 +229,11 @@ func (l *LoadBalancer) insertParameters(dep *appsv1.Deployment) *appsv1.Deployme
 					common.GetProbeCommand(false, "unix:///tmp/health.sock", ""))
 			}
 			container.Env = l.getFeEnvVars(container.Env)
+			// set resource requirements for container (if not found, then values from model
+			// are kept even upon updates, as getReconciledDesiredStatus() overwrites containers)
+			if err := common.SetContainerResourceRequirements(&l.attractor.ObjectMeta, &container); err != nil {
+				l.exec.LogInfo(fmt.Sprintf("attractor %s, %v", l.attractor.ObjectMeta.Name, err))
```
Comment (LionelJouin) [inline controllers/attractor/lb-fe.go:235]: LogError

File: controllers/conduit/proxy.go [hunk 5]
```diff
@@ -139,6 +145,11 @@ func (i *Proxy) insertParameters(init *appsv1.DaemonSet) *appsv1.DaemonSet {
 					common.GetProbeCommand(false, "unix:///tmp/health.sock", ""))
 			}
 			container.Env = i.getEnvVars(container.Env)
+			// set resource requirements for container (if not found, then values from model
+			// are kept even upon updates, as getReconciledDesiredStatus() overwrites containers)
+			if err := common.SetContainerResourceRequirements(&i.conduit.ObjectMeta, &container); err != nil {
+				i.exec.LogInfo(fmt.Sprintf("conduit %s, %v", i.conduit.ObjectMeta.Name, err))
```
Comment (LionelJouin) [inline controllers/conduit/proxy.go:151]: LogError

File: controllers/trench/ipam.go [hunk 6]
```diff
@@ -133,6 +139,11 @@ func (i *IpamStatefulSet) insertParameters(dep *appsv1.StatefulSet) *appsv1.Stat
 					common.GetProbeCommand(false, "unix:///tmp/health.sock", ""))
 			}
 			container.Env = i.getEnvVars(ret.Spec.Template.Spec.Containers[0].Env)
+			// set resource requirements for container (if not found, then values from model
+			// are kept even upon updates, as getReconciledDesiredStatus() overwrites containers)
+			if err := common.SetContainerResourceRequirements(&i.trench.ObjectMeta, &container); err != nil {
+				i.exec.LogInfo(fmt.Sprintf("trench %s, %v", i.trench.ObjectMeta.Name, err))
```
Comment (LionelJouin) [inline controllers/trench/ipam.go:145]: LogError

File: controllers/trench/nsp.go [hunk 7]
```diff
@@ -89,6 +95,11 @@ func (i *NspStatefulSet) insertParameters(init *appsv1.StatefulSet) *appsv1.Stat
 					common.GetProbeCommand(false, "unix:///tmp/health.sock", ""))
 			}
 			container.Env = i.getEnvVars(container.Env)
+			// set resource requirements for container (if not found, then values from model
+			// are kept even upon updates, as getReconciledDesiredStatus() overwrites containers)
+			if err := common.SetContainerResourceRequirements(&i.trench.ObjectMeta, &container); err != nil {
+				i.exec.LogInfo(fmt.Sprintf("trench %s, %v", i.trench.ObjectMeta.Name, err))
```
Comment (LionelJouin) [inline controllers/trench/nsp.go:101]: LogError

File: controllers/common/const.go [hunk 8]
```diff
@@ -49,6 +49,9 @@ const (
 	CMName      = "meridio-configuration"
 
 	NetworkServiceName = "external-vlan"
+
+	ResourceRequirementKey          = "resourcerequirement"
```
Comment (LionelJouin) [inline controllers/common/const.go:53]: @johlj or @tedlean any suggestion?

File: controllers/attractor/nse-vlan.go (hunk 1 above)
Comment (zolug) [reply controllers/attractor/nse-vlan.go:126]: Initially I was using LogError but that also printed a stack dump which I found an overkill and unnecessary spam, thus setlled with LogInfo

File: controllers/common/const.go (hunk 8 above)
Comment (tedlean) [reply controllers/common/const.go]: This is meant to refer to a template that is to inserted as "resources" section in the container manifest, right?
So my suggestion would be "resources-template" or "resource-template"

**Pull request merged**